actions/nmap/NmapAssetScan.py

```diff
@@ -0,0 +1,150 @@
+    service = Entity('Service', alias='service', **svc_kwargs)
+    open_port = ap_pattern.get('openport')
+    merge_pattern = open_port.with_edge(Relationship('is_running')).with_node(service)
+    changes = [(ap_pattern, "merge", merge_pattern)]
+    
+    if svc_kwargs.get("cpe") and svc_kwargs["cpe"] != "unknown":
+        cve_dict = query_scap_for_cve_fuzzy(svc_kwargs["cpe"], conn)
+        for cve_id, details in cve_dict.items():
+            vuln = Entity('Vulnerability', alias='vuln', id=cve_id, source=details[0], criteria=details[1])
+            vuln_pattern = service.with_edge(Relationship('exposes', direction='r')).with_node(vuln)
+            changes.append((service, "merge", vuln_pattern))
+                    
+    return changes
+
+
+class NmapAssetScan(Action):
+    def __init__(self):
+        super().__init__("NmapAssetScan", "T1046", "TA0007", ["loud", "fast"])
+        self.noise = 0.9
+        self.impact = 0
+        self._parsers = {'ftp': ftp_nmap_parser}
+
+    def expected_outcome(self, pattern: Pattern) -> list[str]:
+        return [
+            f"Gain knowledge of network services on {pattern.get('asset').get('ip_address')}. May include additional details for some services, such as indicating if anonymous FTP is supported.",
+        ]
+
+    def get_target_query(self) -> Query:
+        asset = Entity('Asset', alias='asset')
+        query = Query()
+        query.match(asset)
+        query.ret_all()
+        return query
+
+    def function(self, sessions: SessionManager, artefacts: ArtefactManager, pattern: Pattern) -> ActionExecutionResult:
+        asset = pattern.get('asset')
+        uuid = artefacts.placeholder("nmap-asset-scan.xml")
+        out_path = artefacts.get_path(uuid)
+        result = shell("nmap", ["-Pn", "-sT", "-sV", "-O", "-p-", asset.get('ip_address'), "-oX", out_path])
+        result.artefacts["xml_report"] = uuid
+        return result
+
+    def capture_state_change(self, gdb: GraphDB, artefacts: ArtefactManager, pattern: Pattern, output: ActionExecutionResult, conn=None) -> StateChangeSequence:
+        asset = pattern.get('asset')
+        changes: StateChangeSequence = []
+        try:
+            with artefacts.open(output.artefacts["xml_report"], 'r') as f:
+                content = f.read()
+        except KeyError as e:
+            raise ActionExecutionError(e)
+        
+        try:
+            os_version, os_family, os_cpe = nmap_parse_os_version_and_family(content)
+        except ET.ParseError as e:
+            raise ActionExecutionError(e)
+        asset = pattern.get('asset')
+        asset.set('os_version', os_version)
+        asset.set('os_family', os_family)
+        asset.set('cpe', os_cpe)
+
+        changes.append((pattern, 'update', asset))
+        
+        if os_cpe != "unknown":
+            cves = query_scap_for_cve_fuzzy(os_cpe, conn)
+            for cve in cves:
+                vuln = Entity('Vulnerability', alias='vuln', id=cve)
```

Comment:
Should this be:
```suggestion
            for cve_id, details in cve_dict.items():
                vuln = Entity('Vulnerability', alias='vuln', id=cve_id, source=details[0], criteria=details[1])
```

For consistency with the usage above? 🤔